src/core/regimes.rkt

```diff
@@ -42,52 +42,37 @@
            (display ">" port))])
 
 (define (exprs-to-branch-on alts)
-  (define critexpr (critical-subexpression (*start-prog*)))
+  (define critexprs (all-critical-subexpressions (*start-prog*)))
   (define vars (program-variables (alt-program (car alts))))
 
-  (if critexpr
-      (remove-duplicates (cons critexpr vars))
+  (if critexprs
+      (remove-duplicates (append critexprs vars))
       vars))
 
-(define (critical-subexpression prog)
-  (define (loc-children loc subexpr)
-    (map (compose (curry append loc)
-		  list)
-	 (range 1 (length subexpr))))
-  (define (all-equal? items)
-    (if (< (length items) 2) #t
-	(and (equal? (car items) (cadr items)) (all-equal? (cdr items)))))
-  (define (critical-child expr)
-    (let ([var-locs
-	   (let get-vars ([subexpr expr]
-			  [cur-loc '()])
-	     (cond [(list? subexpr)
-		    (append-map get-vars (cdr subexpr)
-				(loc-children cur-loc subexpr))]
-		   [(constant? subexpr)
-		    '()]
-		   [(variable? subexpr)
-		    (list (cons subexpr cur-loc))]))])
-      (cond [(null? var-locs) #f]
-            [(all-equal? (map car var-locs))
-             (caar var-locs)]
-            [#t
-             (let get-subexpr ([subexpr expr] [vlocs var-locs])
-               (cond [(all-equal? (map cadr vlocs))
-                      (get-subexpr (if (= 1 (cadar vlocs)) (cadr subexpr) (caddr subexpr))
-                                   (for/list ([vloc vlocs])
-                                     (cons (car vloc) (cddr vloc))))]
-                     [#t subexpr]))])))
-  (let* ([locs (localize-error prog)])
-    (if (null? locs)
-        #f
-        (let* ([candidate-expr (critical-child (location-get (car locs) prog))]
-               [candidate-prog `(lambda ,(program-variables (*start-prog*)) ,candidate-expr)])
-          (if (and candidate-expr
-                   (for/or ([(pt ex) (in-pcontext (*pcontext*))])
-                     (nan? ((eval-prog candidate-prog 'fl) pt))))
-              #f
-              candidate-expr)))))
+(define (critical-subexpression? expr loc)
```

Comment:
I think with this function, you haven't really thought through cases like `(+ (+ (* a x) z) (* a x))`. Is `(* a x)` a critical subexpression? Your code says "yes", and I do think that that's the way it should be, but given that it can happen, does it make sense to indicate the subexpression by a location? After all, two different locations indicate the same subexpression in this case, and it is the subexpression, not its location, that matters. (Perhaps you see why we use both locations and subexpressions throughout the code.)